### ISSUE TITLE
Help with writing an Adapter

### ISSUE THREAD

**Issue by donpedro** [gulpjs/gulp#2347]
We've been writing a [bunch of plugins](https://gulpetl.com/getting-started/plugins/index) for our new [gulp-etl](https://gulpetl.com/) project, which uses gulp to do data manipulation and automation. The [Writing a plugin](https://github.com/gulpjs/gulp/tree/master/docs/writing-a-plugin) docs have been invaluable, thx!

I'm now writing my first vinyl adapter, and I'm having a harder time getting going there since there doesn't seem to be any analogous "Writing an adapter" documentation. Is there any example of a starter adapter? I'm not having much luck working through the thicket of streaming tools in vinyl-fs.src() to reproduce a basic example.

If not, here's my specific problem: I've got an adapter that works in streaming mode (current code is [here](https://github.com/gulpetl/gulp-etl-tap-mysql/blob/master/src/plugin.ts)), but I'm having a problem getting buffer mode to work. Since there's no callback in .src() I can't figure out how to return an .isBuffer() vinyl file that will later, asynchronously, receive its contents. The file is being evaluated downstream as .isNull() and finishing processing before I can set its contents.

I know this must be a simple issue, but my gulp/streaming ninja skills are still weak :) Can somebody point me in the right direction? If it would help I can post my misguided efforts at coding the buffer mode, but I'm hoping someone can just point me to a good example.

As to buffer mode: we'll be working mainly in streaming mode, but it seems to me that we'd need to provide both modes to be good gulp citizens. Right? Also, once we finish this adapter we'd like to build gulp adapters for S3 and DropBox (discussed [here](https://github.com/gulpjs/vinyl/issues/38), looks like they're both still needed) and we'll want to do them the same way.

**Comment by phated** [comment]
@donpedro I definitely think this is valuable documentation to have.  Currently, vinyl-fs has gotten much more complicated than the original adapter concept (it now includes sourcemaps, symlinks, etc) but I've been trying to think of a way to streamline adapters so you can "plug-in" the reading/writing mechanisms and have all the benefits of other adapters.

I'd love to chat more about this and see if my thoughts are on the right track.

**Comment by donpedro** [comment]
@phated you're tremendously accommodating, thank you!

My initial goals are much more modest; I'm trying to recreate for vinyl adapters the most valuable resource I found for learning to create plugins: [What does a good plugin look like?](https://github.com/gulpjs/gulp/blob/master/docs/writing-a-plugin/guidelines.md#what-does-a-good-plugin-look-like) For instance, here is my original proof of concept .src(), which I wrote to see if I could even get anything to work at all: 
```
import Vinyl = require('vinyl')
import PluginError = require('plugin-error');
const pkginfo = require('pkginfo')(module); // project package.json info into module.exports
const PLUGIN_NAME = module.exports.name;
const from2 = require('from2');
import * as path from 'path'

// dead-simple stream-mode-only gulp vinyl adapter: loads a file as a 
// stream and wraps it in a vinyl file, then returns vinyl file in a readable stream
export function src(filePath:string, optionsIgnoredForNow:any) {
  let fileStream
  let vinylFile
  try {
    fileStream = require('fs').createReadStream(filePath)
    // let asdf = (undefined as any).wontWork // create an error for testing
    vinylFile = new Vinyl({
      base: path.dirname(filePath),    
      path:filePath,
      contents:fileStream
    });
  }
  catch (err) {
    throw new PluginError(PLUGIN_NAME, err);
  }

  return from2.obj([vinylFile])
}
```

It should be simplified further, but the skeleton is here; create a stream using a familiar built-in module (node's fs), demonstrate how to drop it into a Vinyl object and return it using a recommended module (from2). No muss, no fuss, just take it from there for your application.

I got it this far, but this only does streams; I'm struggling with the buffer mode code. If someone can help me add the code for buffer mode that would work with a non-streaming asynchronous resource like [fs.readFile](https://nodejs.org/api/fs.html#fs_fs_readfile_path_options_callback), I think you've got a great starting point that could begin to make it as easy to create vinyl adapters as it is to create plugins. I'm new to the gulp world, but as @contra [wrote 5 years ago](https://medium.com/@contrahacks/gulp-3828e8126466): 

> It seems like people really understood the concept of a “gulp plugin” (a transform stream) but the idea of vinyl adapters never really clicked with many. I’m not sure how to make this catch on...

As a beginning gulp developer, I'd say that this piece (really, the entire [Writing a plugin](https://github.com/gulpjs/gulp/tree/master/docs/writing-a-plugin) doc) which doesn't exist for vinyl adapters is the biggest reason gulp plugin development seems really accessible while building a vinyl adapter looks like a huge mountain to climb.

**Comment by donpedro** [comment]
I found the solution I was looking for, and it was indeed simple. The problem in brief: trying to wrap a streaming resource but still provide a buffer-mode option.

The solution I went with was to simply stream the output through gulp-buffer. Basically, replacing the last two lines of the above code with this:
```
const gulpBuffer = require('gulp-buffer')

...

    let result = from2.obj([vinylFile])
    // buffer mode: stream through gulpBuffer to convert vinylFile to buffer mode (isBuffer(true))
    if (options.buffer !== false) {
      result = result.pipe(gulpBuffer())
    }
  }
```

So with that out of the way it's still a work in progress, but I've now officially published my [first Vinyl adapter](https://www.npmjs.com/package/gulp-etl-mysql-adapter)!

Thanks to @phated and the gulp team for all the work! I'll head over now and see if I can help out a bit with writing some docs.
